test/test_util/test_data.py

```diff
@@ -0,0 +1,632 @@
+
+
+def test_reduce_dataset_invalid_operations():
+    invalid_op = "invalid"
+
+    X = np.asarray([1, 2, 3], dtype=float)
+    with pytest.raises(ValueError) as err:
+        reduce_dataset_size_if_too_large(
+            X=X,
+            y=X,
+            operations=[invalid_op],
+            memory_limit=1,
+            memory_allocation=0.1,
+            is_classification=False
+        )
+
+    expected_err = f"Unknown operation `{invalid_op}`"
+    assert err.value.args[0] == expected_err
+
+
+@parametrize("memory_allocation", [-0.5, 0.0, 1.0, 1.5])
+def test_reduce_dataset_invalid_memory_allocation_float(memory_allocation: float):
+    """
+    Parameters
+    ----------
+    memory_allocation: float
+        An invalid memory allocation as a float
+
+    Expects
+    -------
+    * Should raise a ValueError
+    """
+    with pytest.raises(ValueError, match=r"memory_allocation if float must be in \(0, 1\)"):
+        reduce_dataset_size_if_too_large(
+            X=np.empty(1),
+            y=np.empty(1),
+            memory_limit=100,
+            is_classification=True,
+            memory_allocation=memory_allocation
+        )
+
+
+@parametrize("memory_allocation", [-1, 0, 100, 101])
+def test_reduce_dataset_invalid_memory_allocation_int(memory_allocation: int):
+    """
+    Parameters
+    ----------
+    memory_allocation: float
+        An invalid memory allocation as a int
+
+    Expects
+    -------
+    * Should raise a ValueError
+    """
+    with pytest.raises(ValueError, match=r"memory_allocation if int must be in \(0, memory_limit"):
+        reduce_dataset_size_if_too_large(
+            X=np.empty(1),
+            y=np.empty(1),
+            is_classification=True,
+            memory_limit=100,
+            memory_allocation=memory_allocation
+        )
+
+
+@parametrize("memory_allocation", ["100", {'a': 1}, [100]])
+def test_reduce_dataset_invalid_memory_allocation_type(memory_allocation: Any):
+    """
+    Parameters
+    ----------
+    memory_allocation: Any
+        An invalid memory allocation type
+
+    Expects
+    -------
+    * Should raise a ValueError
+    """
+    with pytest.raises(ValueError, match=r"Unknown type for `memory_allocation`"):
+        reduce_dataset_size_if_too_large(
+            X=np.empty(1),
+            y=np.empty(1),
+            memory_limit=100,
+            is_classification=True,
+            memory_allocation=memory_allocation
+        )
+
+
+@pytest.mark.parametrize(
+    'memory_limit,precision,task',
+    [
+        (memory_limit, precision, task)
+        for task in chain(CLASSIFICATION_TASKS, REGRESSION_TASKS)
+        for precision in (float, np.float32, np.float64, np.float128)
+        for memory_limit in (1, 100)
+    ]
+)
+def test_reduce_dataset_subsampling_explicit_values(memory_limit, precision, task):
+    random_state = 0
+    fixture = {
+        BINARY_CLASSIFICATION: {
+            1: {float: 2621, np.float32: 2621, np.float64: 2621, np.float128: 1310},
```

Comment:
Compared to the "old" test, there's a lot less subsampling going on and I can't see the immediate reason for this. One example is that for regression `np.float64` is no longer subsampled. Do you know why that's the case

Reply:
It seems to mainly be for float64 (and consequently float). I'm not sure why this is either, I'll have a look into this.

Reply:
Here's the old behaviour
```python
Before precision reduction: 0.762939453125MB
After precision reduction: 0.3814697265625MB
Before Subsample reduction: 0.762939453125MB
After Subsample reduction: 0.099945068359375MB
```

Reply:
Here's the old behaviour:
```python
Before precision reduction: 0.762939453125MB
After precision reduction: 0.3814697265625MB
Before Subsample reduction: 0.762939453125MB
After Subsample reduction: 0.0499725341796875MB
```